nifi-daffodil-processors/src/main/java/com/tresys/nifi/processors/AbstractDaffodilProcessor.java

```diff
@@ -140,130 +165,245 @@
+                && externalVariables.equals(otherKey.externalVariables)
+                && tunableMap.equals(otherKey.tunableMap) && configFile.equals(otherKey.configFile);
         }
 
         public int hashCode() {
-          return Objects.hash(dfdlSchema, preCompiled);
+            return Objects.hash(dfdlSchema, validationMode,
+                externalVariables, tunableMap, configFile);
         }
 
-        public boolean equals(Object obj) {
-          if (!(obj instanceof CacheKey)) return false;
-          if (obj == this) return true;
+    }
+
+    public static class DataProcessorSchemaPair {
+        private final DataProcessor dataProcessor;
+        private final RecordSchema schema;
 
-          CacheKey that = (CacheKey)obj;
-          return Objects.equals(this.dfdlSchema, that.dfdlSchema) &&
-                 Objects.equals(this.preCompiled, that.preCompiled);
+        public DataProcessorSchemaPair(DataProcessor dataProcessor, RecordSchema schema) {
+            this.dataProcessor = dataProcessor;
+            this.schema = schema;
         }
-    }
 
-    @Override
-    public Set<Relationship> getRelationships() {
-        return relationships;
+        public DataProcessor getDataProcessor() {
+            return dataProcessor;
+        }
+
+        public RecordSchema getSchema() {
+            return schema;
+        }
     }
 
+    private LoadingCache<CacheKey, DataProcessorSchemaPair> cache;
+
+    /**
+     * Note that the Reader/Writer component always winds up at index 1; we choose which one to add based
+     * on if this is the Parse or Unparse Processor.
+     */
     @Override
     protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> properties = new ArrayList<>(super.getSupportedPropertyDescriptors());
+        properties.addAll(daffodilProperties);
+        if (isUnparse()) {
+            properties.add(1, RECORD_READER);
+        } else {
+            properties.add(1, RECORD_WRITER);
+        }
         return properties;
     }
 
-    private DataProcessor newDaffodilDataProcessor(String dfdlSchema, Boolean preCompiled) throws IOException {
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        if (propertyDescriptorName == null || propertyDescriptorName.replaceAll("\\s+", "").isEmpty()) {
+            return null;
+        } else {
+            boolean isTunable = propertyDescriptorName.startsWith("+");
+            String actualName = isTunable ? propertyDescriptorName.substring(1) : propertyDescriptorName;
+            return new PropertyDescriptor.Builder()
+                .name(actualName)
+                .displayName((isTunable ? "Tunable Variable " : "External Variable ") + actualName)
+                .description("Value to configure a specific " + (isTunable ? "tunable variable" : "external variable") + " when parsing or unparsing")
```

Comment:
I just tested this in NiFi and something weird is going on with tunables using the NiFi GUI. When I add a tunable starting with +, it doesn't show up and I so I can't edit its value. If I then add a variable, it adds correctly, but there's a space where the tunable was added. So seems like the tunable is added, but something prevents it from showing, likely a bug in NiFi.

 I'm wondering if there is an issue with changing the "actualName"? I made this modification and it at least allows me to add and edit the tunable.

```suggestion
            String displayName = isTunable ? propertyDescriptorName.substring(1) : propertyDescriptorName;
            return new PropertyDescriptor.Builder()
                .name(propertyDescriptorName)
                .displayName((isTunable ? "Tunable: " : "External Variable: ") + displayName)
                .description("Value to configure the " + displayName + " " + (isTunable ? "tunable" : "external variable") + " when parsing or unparsing")
```

NiFi doesn't seem to show the display name until after you close and reopen the configure processor properties, but it does seem to fix the GUI issues.